File: historian_v3/pkg/historian/repair.go
package historian

import (
	"os"
	"path/filepath"
)

func RepairStorageTail(cfg Config) error {
	rawRoot := filepath.Join(cfg.Storage.DataDir, "raw")
	_ = filepath.Walk(rawRoot, func(path string, info os.FileInfo, err error) error {
		if err != nil || info == nil || info.IsDir() || filepath.Ext(path) != ".seg" {
			return nil
		}
		b, err := os.ReadFile(path)
		if err != nil {
			return nil
		}
		off := 0
		for off < len(b) {
			_, n, ok := DecodeSegmentRecord(b, off)
			if !ok {
				break
			}
			off += n
		}
		if off < len(b) {
			if f, err := os.OpenFile(path, os.O_WRONLY, 0o644); err == nil {
				_ = f.Truncate(int64(off))
				_ = f.Close()
			}
		}
		return nil
	})
	idxRoot := filepath.Join(cfg.Storage.DataDir, "index")
	_ = filepath.Walk(idxRoot, func(path string, info os.FileInfo, err error) error {
		if err != nil || info == nil || info.IsDir() || filepath.Ext(path) != ".idx" {
			return nil
		}
		sz := int(info.Size())
		good := (sz / BlockIndexEntrySize) * BlockIndexEntrySize
		if good < sz {
			if f, err := os.OpenFile(path, os.O_WRONLY, 0o644); err == nil {
				_ = f.Truncate(int64(good))
				_ = f.Close()
			}
		}
		return nil
	})
	return nil
}

File: historian_v2/src/storage/repair.rs
use anyhow::Result;
use std::path::{Path, PathBuf};

use crate::config::Config;
use crate::storage::codec::{decode_segment_record, BLOCK_INDEX_ENTRY_SIZE};

async fn walk_files(root: &Path, ext: &str) -> Result<Vec<PathBuf>> {
    let mut out = Vec::new();
    if !root.exists() {
        return Ok(out);
    }
    let mut stack = vec![root.to_path_buf()];
    while let Some(cur) = stack.pop() {
        let mut rd = tokio::fs::read_dir(&cur).await?;
        while let Some(entry) = rd.next_entry().await? {
            let p = entry.path();
            let meta = entry.metadata().await?;
            if meta.is_dir() {
                stack.push(p);
            } else if p.extension().map(|v| v == ext).unwrap_or(false) {
                out.push(p);
            }
        }
    }
    Ok(out)
}

pub async fn repair_storage_tail(cfg: &Config) -> Result<()> {
    let raw_root = Path::new(&cfg.storage.data_dir).join("raw");
    let idx_root = Path::new(&cfg.storage.data_dir).join("index");
    let segs = walk_files(&raw_root, "seg").await?;
    for seg in segs {
        let data = tokio::fs::read(&seg).await?;
        let mut off = 0usize;
        while off < data.len() {
            if let Some((_, n)) = decode_segment_record(&data, off) {
                off += n;
            } else {
                break;
            }
        }
        if off < data.len() {
            tokio::fs::OpenOptions::new()
                .write(true)
                .open(&seg)
                .await?
                .set_len(off as u64)
                .await?;
        }
    }

    let idxs = walk_files(&idx_root, "idx").await?;
    for idx in idxs {
        let meta = tokio::fs::metadata(&idx).await?;
        let good = (meta.len() as usize / BLOCK_INDEX_ENTRY_SIZE) * BLOCK_INDEX_ENTRY_SIZE;
        if good < meta.len() as usize {
            tokio::fs::OpenOptions::new()
                .write(true)
                .open(&idx)
                .await?
                .set_len(good as u64)
                .await?;
        }
    }
    Ok(())
}
